### PR TITLE
Allow picking the date field to filter by (uploaded, taken, modified)

### PR DESCRIPTION
**Pull request by theefer**
image

This is done by adding a new `dateField` and `{taken,uploaded}{Since,Until}` parameters to the state (kahuna URI), and sending them along to the media-api using the already supported `{taken,uploaded}{Since,Until}` parameters.

cc @blishen

### CODE REVIEW

**Comment by paperboyo** [comment]
Cool! Do we know what's the percentage of images with 'dateTaken'? If you sort by 'dateTaken' what happens to the images that do not have it (does it fallback to 'modified', and vice versa)? Does any user activity in Grid change 'modified'?

**Comment by theefer** [comment]
About 2M images have dateTaken  out of the 3.5M images in the db. Images without one are simply not returned when filtering by dateTaken, which kind of makes sense, but is worth noting. Have to get numbers on the dateTaken coverage for recent images... 

All user edits on metadata, usage crops etc bump last modified.

File: kahuna/public/js/search/index.js [hunk 1]
```diff
@@ -129,7 +129,8 @@ search.config(['$stateProvider', '$urlMatcherFactoryProvider',
     });
 
     $stateProvider.state('search.results', {
-        url: 'search?{query:Query}&ids&since&nonFree&uploadedBy&until&orderBy',
+        url: 'search?{query:Query}&ids&since&nonFree&uploadedBy&until&orderBy' +
+             '&dateField&takenSince&takenUntil&modifiedSince&modifiedUntil',
```
Comment (jamesgorrie) [inline kahuna/public/js/search/index.js]: I know you can use the date type, although I haven't done it before. i.e. `{modifiedUntil:date}`.
That applies to `boolean`s and `int`s too.

**Comment by paperboyo** [comment]
> About 2M images have dateTaken out of the 3.5M images in the db

That's a much better ratio than I thought. If indeed most of new images have this field, then



> Images without one are simply not returned when filtering by dateTaken

wouldn't present a problem. :-)

> All user edits on metadata, usage crops etc bump last modified.

Super-cool! Thanks for explanation.

File: kahuna/public/js/search/index.js (hunk 1 above)
Comment (theefer) [inline kahuna/public/js/search/index.js]: Good to know thanks. I think I'm going to leave as a String for now because that's how everything else works and I don't really want to refactor so much more unrelated stuff as part of this change..

**Comment by akash1810** [comment]
:+1:

File: kahuna/public/js/search/index.js (hunk 1 above)
Comment (jamesgorrie) [inline kahuna/public/js/search/index.js]: Yea, I only found it out the other day and thought it was quite nice.

Comment (akash1810) [inline kahuna/public/js/search/index.js]: Is that what's going on with `{query:Query}` too? If so, its not restricted to native types!

Comment (jamesgorrie) [inline kahuna/public/js/search/index.js]: Nope, you can write your own types, but some are native.
I can't find an exhaustive list though.

Comment (jamesgorrie) [inline kahuna/public/js/search/index.js]: [More on custom types though](http://angular-ui.github.io/ui-router/site/#/api/ui.router.util.$urlMatcherFactory#methods_type).

Comment (jamesgorrie) [inline kahuna/public/js/search/index.js]: [And here we are](https://github.com/angular-ui/ui-router/blob/master/src/params/paramTypes.ts#L21)
